File: 2020/5. Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        int start = 0, end = 0, max = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expand(s, i, i);
            int len2 = expand(s, i, i + 1);
            if (len1 > max) {
                max = len1;
                start = i - len1 / 2;
                end = i + len1 / 2;
            } 
            if (len2 > max) {
                max = len2;
                start = i - (len2 - 1) / 2;
                end = i + len2 / 2;
            }
        }
        return s.substring(start, end + 1);
    }
    
    private int expand(String s, int left, int right) {
        while (left > -1 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;            
        }
        return right - left - 1;
    }
}

File: 2024/5. Longest Palindromic Substring.cpp
class Solution {
public:
    void expand(string s, int l, int r, int& max_len, int& max_st) {
        while (l >= 0 && r < s.length() && s[l] == s[r]) {
            if (r - l + 1 > max_len) {
                max_st = l;
                max_len = r - l + 1;
            }
            l--;
            r++;
        }
    }
    string longestPalindrome(string s) {
        int max_len = 0, max_st = 0;
        for (int i = 0; i < s.length(); i++) {
            expand(s, i, i, max_len, max_st);
            expand(s, i, i + 1, max_len, max_st);
        }
        return s.substr(max_st, max_len);
    }
};